woocommerce/woocommerce
# POC: Product form templates via block hooks API

### Changes proposed in this Pull Request:

This PR is a POC demonstrating how we can realign with Gutenberg's templating APIs and provide extensibility.  It does the following:

* Converts a sample product template into comment-delimited template format
* Hooks the product templates into the block hooks and adds them to the REST endpoint, potentially allowing templates to be customized by themes or users
* Extends the Woo template abstraction to include product type and default data and connect this to the actual Gutenberg template
* Consumes templates from the WP template parts endpoint
* This PR intentionally doesn't block off form templates from the site editor to showcase its visibility there
<img width="349" alt="Screenshot 2024-02-05 at 7 49 32 PM" src="https://github.com/woocommerce/woocommerce/assets/10561050/7e716f10-f3c2-4051-91b9-18b539022e03">
<img width="837" alt="Screenshot 2024-02-05 at 7 25 16 PM" src="https://github.com/woocommerce/woocommerce/assets/10561050/932448bc-f13e-41d9-a6fd-8b2f0d04faf7">



### Discussion

After this, my initial impressions are positive and overall I like the idea of moving towards this direction.  It affords us quite a few benefits:

* The HTML comment delimited templates feel more natural to write and are easier.  We can still use template parts to quickly compose templates together
* We don't need to maintain complex template APIs
* If we decide to open up to the idea of custom product templates or allow themes to provide templates for products, this is pretty trivial to do with this concept

However, there are some missing features upstream that make this currently a challenge.  I'll post more about this tomorrow, but here are some highlights:

* Block hooks don't support adding attributes (huge blocker)
* We still need some addtiional data around each individual block that doesn't currently fit into the block abstraction (e.g., hiding/disabling blocks).  This could be worked around with some additional wrappers, but I'm hesitant to create a separate API again.
* Anchoring vs order/priority positioning of blocks is still debateable.

Finally, if we do decide to go through with this, we have a good bit of overhaul we need to handle in this repo:

* Deprecation of existing API methods and temporarily adapting them to wrap around block hooks
* Migrating templates to template parts in comment delimited format.  I think template parts may be more fitting than templates since the form itself only represents a piece of the page.
* Updating client-side template consumption to use all the new templates and template data

### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Make a GET request to `/wp-json/wp/v2/template-parts`
2. Note that `woocommerce/woocommerce//simple-form` is in the response
3. Note the title and description match those in the `SimpleProductTemplate` class
4. Note the additional data (product types and default data) are also included in the response
5. Make a GET request to `/wp-json/wp/v2/template-parts?area=product-form` and note that only the product form templates are included in the response
6. Navigate to the new product editing experience
7. Note that the 2 product tabs, basic details section, and price field are shown
8. To demonstrate extensibility, add the following snippet to a plugin:
```php
function woo_add_sku_field( $hooked_blocks, $position, $anchor_block, $context ) {
	if ( 
		'woocommerce/product-name-field' === $anchor_block &&
		'after' === $position &&
		'product-form' === $context->area
	) {
		$hooked_blocks[] = 'woocommerce/product-sku-field';
	}

	return $hooked_blocks;
}
add_filter( 'hooked_block_types', 'woo_add_sku_field', 10, 4 );
```
9. Load the product editor once more and note the sku field right after the price field.
10. Navigate to Appearance -> Editor -> Patterns
11. Note the "Product Form" section under "Template Parts"
12. Under that section you should see "Simple Product Form Template" (functionality will not yet work in this PR as blocks have not been loaded on this page)



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
